File: frontend/src/api/series.ts
import client from './client';

export interface SeriesProjectItem {
  id: string;
  name: string;
  genre: string | null;
  status: string;
  sort_order: number;
  cover_image: string | null;
}

export interface Series {
  id: string;
  name: string;
  description: string | null;
  projects: SeriesProjectItem[];
  created_at: string;
  updated_at: string;
}

export interface SeriesCreate {
  name: string;
  description?: string;
  project_ids?: string[];
}

export interface SeriesUpdate {
  name?: string;
  description?: string;
}

export const seriesApi = {
  list: () => client.get<Series[]>('/series'),
  get: (id: string) => client.get<Series>(`/series/${id}`),
  create: (data: SeriesCreate) => client.post<Series>('/series', data),
  update: (id: string, data: SeriesUpdate) => client.put<Series>(`/series/${id}`, data),
  delete: (id: string) => client.delete(`/series/${id}`),
  addProject: (seriesId: string, projectId: string) =>
    client.post(`/series/${seriesId}/projects/${projectId}`),
  removeProject: (seriesId: string, projectId: string) =>
    client.delete(`/series/${seriesId}/projects/${projectId}`),
  reorder: (seriesId: string, projectIds: string[]) =>
    client.put(`/series/${seriesId}/reorder`, { project_ids: projectIds }),
};

File: backend/app/schemas/series.py
import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class SeriesCreate(BaseModel):
    name: str = Field(..., max_length=200)
    description: Optional[str] = None
    project_ids: list[uuid.UUID] = Field(default_factory=list, max_length=20)


class SeriesUpdate(BaseModel):
    name: Optional[str] = Field(default=None, max_length=200)
    description: Optional[str] = None


class SeriesProjectReorder(BaseModel):
    project_ids: list[uuid.UUID]


class SeriesProjectItem(BaseModel):
    id: uuid.UUID
    name: str
    genre: Optional[str]
    status: str
    sort_order: int
    cover_image: Optional[str]

    model_config = {"from_attributes": True}


class SeriesResponse(BaseModel):
    id: uuid.UUID
    name: str
    description: Optional[str]
    projects: list[SeriesProjectItem] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
